src/webgpu/web_platform/copyToTexture/canvas.spec.ts

```diff
@@ -214,17 +317,131 @@ g.test('copy_contents_from_canvas')
 
     // Construct expected value for different dst color format
     const dstBytesPerPixel = kTextureFormatInfo[dstColorFormat].bytesPerBlock;
+    const format: RegularTextureFormat =
+      dstColorFormat === 'rgba8unorm-srgb' || dstColorFormat === 'bgra8unorm-srgb'
+        ? dstColorFormat === 'rgba8unorm-srgb'
+          ? 'rgba8unorm'
+          : 'bgra8unorm'
+        : dstColorFormat;
+    dstColorFormat;
+
+    // For 2d canvas, get expected pixels with getImageData(), which returns origin
+    // values(not applied premultipy alpha) always.
     const expectedPixels = t.getExpectedPixels({
       context: canvasContext,
       width,
       height,
+      format,
+      contextType: '2d',
+      srcPremultiplied: false,
+      dstPremultiplied: premultipliedAlpha,
+    });
+
+    t.doTestAndCheckResult(
+      { source: canvas, origin: { x: 0, y: 0 } },
+      {
+        texture: dst,
+        origin: { x: 0, y: 0 },
+        colorSpace: 'srgb',
+        premultipliedAlpha,
+      },
+      { width: canvas.width, height: canvas.height, depthOrArrayLayers: 1 },
+      dstBytesPerPixel,
+      expectedPixels
+    );
+  });
+
+g.test('copy_contents_from_gl_context_canvas')
+  .desc(
+    `
+  Test HTMLCanvasElement and OffscreenCanvas with webgl/webgl2 context
+  can be copied to WebGPU texture correctly.
+
+  It creates HTMLCanvasElement/OffscreenCanvas with webgl'/'webgl2'.
+  Use fillRect(2d context) or stencil + clear (gl context) to rendering
+  red rect for top-left, green rect for top-right, blue rect for bottom-left
+  and black for bottom-right.
+  Then call copyExternalImageToTexture() to do a full copy to the 0 mipLevel
+  of dst texture, and read the contents out to compare with the canvas contents.
+
+  The tests covers:
+  - Valid canvas type
+  - Valid webgl/webgl2 context type
+  - TODO: color space tests need to be added
+
+  And the expected results are all passed.
+  `
+  )
+  .params(u =>
+    u
+      .combine('canvasType', ['onscreen', 'offscreen'] as const)
+      .combine('contextName', ['webgl', 'webgl2'] as const)
+      .combine('dstColorFormat', kValidTextureFormatsForCopyIB2T)
+      .combine('webglCanvasPremultipliedAlpha', [true, false])
+      .combine('premultipliedAlpha', [true, false])
```

Comment:
nit: I think `srcPremultiplied` and `dstPremultiplied`, as you already use in getExpectedPixels, are more self-explanatory than these.